src/desktop/credentials/DesktopNativeCredentialsFacade.ts

```diff
@@ -1,35 +1,136 @@
+				// if we actually have pin-encrypted credentials and just lost the salt, we
+				// will get a decryption failure and delete the credentials further up the stack.
+				return data
+			}
+			try {
+				data = this.crypto.aesDecryptBytes(appPinKey, data)
+			} catch (e) {
+				if (e instanceof CryptoError) {
+					const nativeFacade = await this.getCurrentCommonNativeFacade()
+					nativeFacade.showAlertDialog("invalidPassword_msg")
+					throw new CancelledError("app pin verification failed")
+				} else {
+					throw e
+				}
+			}
+			return data
+		} else {
+			// our mode is not app pin, so the app pin should not be set.
+			// this is technically not required because we're deleting it when
+			// encrypting with another mode, but things might have gone wrong.
+			await this.conf.setVar(DesktopConfigKey.appPinSalt, null)
+			return data
+		}
+	}
+
+	private async maybeEncryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		if (encryptionMode === CredentialEncryptionMode.APP_PIN) {
+			const appPinKey = await this.resolveKeyFromAppPin(true)
+			if (appPinKey == null) {
+				throw new ProgrammingError("we should have generated a new key or thrown a CancelledErrorq")
+			}
+			data = this.crypto.aesEncryptBytes(appPinKey, data)
+			return data
+		} else {
+			// our mode is not app pin, so the app pin salt should not be set
+			await this.conf.setVar(DesktopConfigKey.appPinSalt, null)
+			return data
+		}
+	}
+
+	/**
+	 * if there is a salt stored, use it and a password prompt to derive the app pin key.
+	 * if there isn't, ask for a new password, generate a salt & store it, then derive the key.
+	 * @param generate whether to generate a new key if no salt is found
+	 * @return the derived 256 bit key or null if none is found and generate is false
+	 */
+	private async resolveKeyFromAppPin(generate: boolean): Promise<Aes256Key | null> {
+		const storedAppPinSaltB64 = await this.conf.getVar(DesktopConfigKey.appPinSalt)
+		const commonNativeFacade = await this.getCurrentCommonNativeFacade()
+		if (storedAppPinSaltB64 == null) {
+			if (!generate) {
+				return null
+			}
+			const newSalt = this.crypto.randomBytes(KEY_LENGTH_BYTES_AES_256)
+			const newPw = await commonNativeFacade.promptForNewPassword(this.lang.get("credentialsEncryptionModeAppPin_label"), null)
+			const newAppPinSaltB64 = uint8ArrayToBase64(newSalt)
+			await this.conf.setVar(DesktopConfigKey.appPinSalt, newAppPinSaltB64)
+			return await this.argon2idFacade.generateKeyFromPassphrase(newPw, newSalt)
+		} else {
+			const pw = await commonNativeFacade.promptForPassword(this.lang.get("credentialsEncryptionModeAppPin_label"))
+			const salt = base64ToUint8Array(storedAppPinSaltB64)
+			return await this.argon2idFacade.generateKeyFromPassphrase(pw, salt)
+		}
+	}
+
+	async getSupportedEncryptionModes(): Promise<Array<DesktopCredentialsMode>> {
+		return SUPPORTED_MODES.slice()
```

Comment:
if you would freeze them you probably don't have to slice but actually doesn't matter